Add a new look-and-feel for the web UI.

ewanmellor opened the pull request:
This adds Roboto / Roboto Mono as web fonts, and adds highlight colors
for table heading and the nav menu based on the OCaml logo.  It adds the
logo itself to the nav bar and the favicon.
    
This adds a static module, which serves static files from a specified
docroot.  Hook this in to the routes under /images/*.
    
Lots of padding and spacing adjustments too.

Set the diagram node color (i.e. the box outline color) to match the fill.
This makes the boxes less heavy on the page, because they no longer have
the black border.
    
Signed-off-by: Ewan Mellor <ewan@tarides.com>

talex5 commented on lib_web/static.ml:
```diff
@@ -0,0 +1,21 @@
+let serve_file ~docroot uri =
+  let fname = Utils.Server.resolve_local_file ~docroot ~uri in
```
I think it would be better to list the resources explicitly here. In theory, `resolve_local_file` will deal with `..` and similar, but I'm not sure relying on that is a good idea. Especially on Windows, where someone might ask to read `CON` or something.

ewanmellor replied:
`resolve_local_file` does indeed handle `..` and similar (I wrote the unit tests for that recently).

The next line in that code is passing the resolved filename to `Utils.Server.respond_file`, which is `Cohttp_lwt_unix.Server.respond_file`.  That function checks that the given filename resolves to a real file (`Unix.st_kind = S_REG`) so it will not serve special files.

Also, this is not serving arbitrary files even from the specified docroot, because it is under the control of `Current_web.routes`, which restricts the URL to `/images/*`.  I think that is sufficiently constrained.

talex5 commented on lib_web/context.ml:
```diff
@@ -2,6 +2,14 @@ open Lwt.Infix
 
 let cookie_key = "__session"
 
+let img_ocaml_logo = "/images/ocaml%20logo.png"
```
Why should OCurrent pipelines have an OCaml logo? Performing CI for OCaml projects isn't the only use for pipelines.

ewanmellor replied:
I have renamed this to dashboard-logo.png.  Pipelines have full control over this image (because they control the docroot that images are served from) so with a generic name I think that all use cases are covered.

talex5 commented on lib_web/context.ml:
```diff
@@ -72,15 +80,19 @@ let template t contents =
       (body [
           nav [
+            img ~alt:"" ~src:img_ocaml_logo ~a:[ a_height 29 ] ();
+            div [code [txt site.name]];
```
The example apps do use the shell command as their name, but a pipeline's name is not "code" in general. Maybe just put a class on this and let the CSS decide what to do?

ewanmellor replied:
I've changed this to a CSS class.

talex5 commented on lib_web/context.ml:
```diff
@@ -72,15 +80,19 @@ let template t contents =
+            img ~alt:"" ~src:img_ocaml_logo ~a:[ a_height 29 ] ();
+            div [code [txt site.name]];
             ul (
-              li [a ~a:[a_href "/"] [txt site.name]] ::
```
Normally, sites make the top-left logo/site-name go to the home page. If having two links is a problem, I suggest getting rid of the "Home" link instead.

ewanmellor replied:
I've put this link back.

talex5 reviewed (commented):
Thanks. The styling does need some work. The new design doesn't seem to work well with narrow windows, though.

Before (from ocaml-ci, running a slightly older version):

![Screenshot_2021-04-13 ocaml-ci](https://user-images.githubusercontent.com/554131/114548089-34642c00-9c57-11eb-87ec-463ee90ef9ce.png)

After:

![Screenshot_2021-04-13 docker_build_local](https://user-images.githubusercontent.com/554131/114548082-2f9f7800-9c57-11eb-866a-2e7a7b5d42fc.png)

talex5 commented on lib_term/dot.ml:
```diff
@@ -30,7 +30,9 @@ let node f ?style ?shape ?bg ?url ?tooltip i label =
     "label", Some label;
+    "color", bg;
     "fillcolor", bg;
+    "fontname", Some "Roboto, sans-serif";
```
It might be better to set the font globally, rather than duplicate it on every node. e.g. https://stackoverflow.com/a/10880578/50926

ewanmellor replied:
I have done this.  There is now a new changeset in this PR, which adds a helper for writing the digraph stanza.  The fontname is then added at that point.

ewanmellor commented on lib_web/current_web.mli:
```diff
@@ -100,7 +100,7 @@ module Resource : sig
 end
 
-val routes : Current.Engine.t -> Resource.t Routes.route list
+val routes : ?docroot:string -> Current.Engine.t -> Resource.t Routes.route list
```
Added.

ewanmellor commented on lib_web/style.ml:
```diff
@@ -47,19 +106,27 @@ let css = {|
-    padding-left: 0.5em;
-    padding-right: 0.5em;
+    padding: 2ex 0.75em;
+    transition: background 300ms, color 300ms;
```
I've turned them into a kind of underline.  See if you like that better.

ewanmellor commented:
> Thanks. The styling does need some work. The new design doesn't seem to work well with narrow windows, though.
> 
> Before (from ocaml-ci, running a slightly older version):
> 
> ![Screenshot_2021-04-13 ocaml-ci](https://user-images.githubusercontent.com/554131/114548089-34642c00-9c57-11eb-87ec-463ee90ef9ce.png)
> 
> After:
> 
> ![Screenshot_2021-04-13 docker_build_local](https://user-images.githubusercontent.com/554131/114548082-2f9f7800-9c57-11eb-866a-2e7a7b5d42fc.png)
> 
> Also, CI is failing.

I have rebased, fixed the failing tests, and addressed all the other comments inline.

I have addressed the narrow-screen issue above by making the `query-form` class contents wrap at medium sizes, and then go to a single column form on mobile screens.

This is now ready for review again.

tmcgilchrist commented on lib_web/static.ml:
```diff
@@ -0,0 +1,21 @@
+
+  Note that we ignore the given path, and we use the Context.uri instead.
+  The routes package has added a wildcard option
+  (https://github.com/anuragsoni/routes/pull/118) but this has not been
```
This has been released as 1.0.0

tmcgilchrist commented on lib_term/s.ml:
```diff
@@ -109,21 +109,21 @@ module type TERM = sig
       each element in [x], once all elements of [x] have successfully completed. *)
 
-  val option_map : ('a t -> 'b t) -> 'a option t -> 'b option t
+  val option_map : ?label:string -> ('a t -> 'b t) -> 'a option t -> 'b option t
```
Could you add docs for `?label`?
